modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php

```diff
@@ -0,0 +1,157 @@
+<?php
+/**
+ * WordPress.com Block Editor
+ * Allow new block editor posts to be composed on WordPress.com.
+ * This is auto-loaded as of Jetpack v7.4 for sites connected to WordPress.com only.
+ *
+ * @package Jetpack
+ */
+
+/**
+ * WordPress.com Block editor for Jetpack
+ */
+class Jetpack_WPCOM_Block_Editor {
+	/**
+	 * ID of the user who signed the nonce.
+	 *
+	 * @var int
+	 */
+	private $nonce_user_id;
+
+	/**
+	 * Singleton
+	 */
+	public static function init() {
+		static $instance = false;
+
+		if ( ! $instance ) {
+			$instance = new Jetpack_WPCOM_Block_Editor();
+		}
+
+		return $instance;
+	}
+
+	/**
+	 * Jetpack_WPCOM_Block_Editor constructor.
+	 */
+	private function __construct() {
+		add_action( 'admin_init', array( $this, 'disable_send_frame_options_header' ), 9 );
+		add_filter( 'admin_body_class', array( $this, 'add_iframed_body_class' ) );
+	}
+
+	/**
+	 * Prevents frame options header from firing if this is a whitelisted iframe request.
+	 */
+	public function disable_send_frame_options_header() {
+		if ( $this->framing_allowed() ) {
+			remove_action( 'admin_init', 'send_frame_options_header' );
+		}
+	}
+
+	/**
+	 * Adds custom admin body class if this is a whitelisted iframe request.
+	 *
+	 * @param string $classes Admin body classes.
+	 * @return string
+	 */
+	public function add_iframed_body_class( $classes ) {
+		if ( $this->framing_allowed() ) {
+			$classes .= ' is-iframed ';
+		}
+
+		return $classes;
+	}
+
+	/**
+	 * Checks whether this is a whitelisted iframe request.
+	 *
+	 * @return bool
+	 */
+	public function framing_allowed() {
+		if ( empty( $_GET['frame-nonce'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+			return false;
+		}
+
+		$verified = $this->verify_frame_nonce(
+			$_GET['frame-nonce'],  // phpcs:ignore WordPress.Security.NonceVerification
+			'frame-' . Jetpack_Options::get_option( 'id' )
+		);
+
+		if ( is_wp_error( $verified ) ) {
+			wp_die( esc_html( $verified ) );
```

Comment:
Let's also make sure it dies with a non-500 status code. Something like:

```
wp_die( esc_html( $verified->get_error_message() ), '', array( 'response' => 400 ) );
```

Or add the status code straight to the `WP_Error` object, and pass that object directly to `wp_die()`:
```
$error = new WP_Error( 'foo', 'error message', array( 'status' => 400 ) );
...
wp_die( $error );
```

Reply:
#12135 has those fixes, I'll merge them over